### ISSUE TITLE
Database Cache Support

### ISSUE THREAD

**Issue by dbpolito** [laravel/ideas#638]
Not sure if this was already discussed previously i sounds a lovely feature to the framework.

What people usually do when you need to cache a query is:

````php
$users = Cache::remember('users', 60, function() {
    return User::get();
});
````

What i would love to do is:

````php
$users = User::remember(60)->get();
// or
$users = User::cache(60)->get();
````

What this basically does is turns the query into md5 as cache key and use the Cache layer we already have on the framework.

I can definitely work on a PR for this, just wanted to confirm if there is any downside of doing this.

**Comment by sebastiandedeyne** [comment]
This used to exist back in Laravel 4 (https://laravel.com/docs/4.2/queries#caching-queries)

Don't know why it was removed, but could be that there was a good reason for it.

**Comment by dbpolito** [comment]
Yeah, i remember that, good catch... forgot to mention :+1:

**Comment by decadence** [comment]
How do you clear this value from cache if you don't specify key manually?

**Comment by dbpolito** [comment]
Well we could have a second param on the method to pass a custom key name:

````php
$users = User::remember(60, 'users')->get();
````

Or make use of tags, as it had on 4.2.

**Comment by drbyte** [comment]
Here's where it was removed: https://github.com/laravel/framework/commit/ea6b9589514b0139508b35166b63e9fbc045c285

If you scroll down to the comments you'll see a few people were offering alternate approaches.  I don't see where a PR was made to bring it back differently though.

**Comment by sisve** [comment]
Caching is hard in general, and very hard to implement so that it works for everyone on a framework level.

What will `$list = Model::remember(60)->where('updated_at', '>', Carbon::now()->subSeconds(60))->get();` remember?

Should the entire invocation be cached for 60 seconds so that the rendered $list only updates once per minute? Should every second cause a new query since we're using second precision? How about milli/nano-second precision that PHP 7 introduced for the DateTime object?

Imagine how this would play out with a soft-deletable-implementation that checked `((deleted_at IS NULL) OR (deleted_at > :now))`

Sure, we can ignore that. Can we say that an static php query can be cached without issues? No, we can issue an identical query twice, without changing the database at all, and still get different results. `... ORDER BY rand()`

TL;DR: Caching is hard. We should not make it so easy that people add a ->remember() call on every query as a routine. Instead, Cache::remember forces them to build the keys and figuring out the actual dependencies of the query themselves.
